### PR TITLE
ctpdev: Digits to Raw data dev

### PR DESCRIPTION
**Pull request by lietava**
@shahor02 : please, have a look.

### CODE REVIEW

**Comment by lietava** [comment]
oops, forgot cling. Doing now.

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: note that the DetID has a method getName, i.e. instead of storing the name you can simply call
`DetID::getName(detID)``.
Alternatively, you can use DetID directly as a data member, instead of DetID::ID.

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: why do you need both detectorsMask and `std::vector<o2::detectors::DetID> detectors` ? Don't they encode the same detectors list?

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: `const std::string& ctpconfiguartion`

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h:43]: is this detectors input mask? If yes, you could probably use DetID::mask_t

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Digits.h
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Digits.h]: for the record: the ROOT schema evolution does not support change of N in the `bitset<N>`.
Are these `CTP_NINPUTS` and `CTP_NCLASSES` constants which never change?

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx]: see my comment about the DetID::getName

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx]: don't use `std::cout`, use `LOG(INFO)` instead.

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx:92]: `const std::string& line`

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx]: add `{}`

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx]: better throw exception: `throw std::runtime_error(fmt::format("INPUT line {}: Detector {} does not exist", line, inp.detName);`

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx]: `LOG(ERROR)`

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [inline DataFormats/Detectors/CTP/src/Configuration.cxx]: same here, also below

File: Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h
Comment (shahor02) [inline Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h]: this will be a simple executable, why do you need root dictionary?

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (shahor02) [inline Detectors/CTP/simulation/src/Digits2Raw.cxx]: You should not have the same `mLinkID==0` on the same CRU. What is the point of this mLinkID at all if you know that you have 2 links?

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (shahor02) [inline Detectors/CTP/simulation/src/Digits2Raw.cxx]: this seems to be a copy/paste error from the example, remove it.

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (lietava) [reply DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h:43]: No, inputMask is position of CTP input in CTP hardware

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Digits.h
Comment (lietava) [reply DataFormats/Detectors/CTP/include/DataFormatsCTP/Digits.h]: No, unless we change hardware.

File: Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h
Comment (lietava) [reply Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h]: This is a class which would be used by simple executable. Do I need root dictionary ?

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (lietava) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: Fixed: mLinkID -> ilink

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (lietava) [reply DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: added full DetID to CTPInput

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (lietava) [reply DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: ok, keeping only DEtID vector

**Comment by lietava** [comment]
@shahor02 I tried to address your comments. Please, have a look.

File: Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h
Comment (shahor02) [reply Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h]: no

**Review by lietava — COMMENTED** [review]
created by mistake

File: Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h
Comment (lietava) [reply Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h]: @shahor02  I removed Digits2Raw from LinkDef. Also there is no ClassImp in cxx. But when I remove also line ClassDefNV(Digits2Raw, 1); from header it does not compile,  so I left it. Is it ok ?

Comment (shahor02) [reply Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h]: I suspect this happens because you did not suppress https://github.com/AliceO2Group/AliceO2/blob/da03f266d029beb6119a6742a3abd18ee08aef34/Detectors/CTP/simulation/CMakeLists.txt#L20 after suppressing  `ClassDefNV(Digits2Raw, 1);`

Comment (lietava) [reply Detectors/CTP/simulation/include/CTPSimulation/Digits2Raw.h]: After removing it still complains, please, see below.
Do I need to clean ?
R.
Detectors/CTP/simulation/G__O2CTPSimulation.cxx: In function 'ROOT::TGenericClassInfo* ROOT::GenerateInitInstanceLocal(const o2::ctp::Digits2Raw*)':
Detectors/CTP/simulation/G__O2CTPSimulation.cxx:91:65: error: 'Class_Version' is not a member of 'o2::ctp::Digits2Raw'
   91 |          instance("o2::ctp::Digits2Raw", ::o2::ctp::Digits2Raw::Class_Version(), "", 134,
      |                                                                 ^~~~~~~~~~~~~
Detectors/CTP/simulation/G__O2CTPSimulation.cxx:93:43: error: 'Dictionary' is not a member of 'o2::ctp::Digits2Raw'
   93 |                   &::o2::ctp::Digits2Raw::Dictionary, isa_proxy, 4,
      |                                           ^~~~~~~~~~

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: in principle, instead of the std::vector<o2::detectors::DetID> you could simply use DetID::mask_t

File: DataFormats/Detectors/CTP/src/Configuration.cxx
Comment (shahor02) [reply DataFormats/Detectors/CTP/src/Configuration.cxx:92]: missed this?

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (shahor02) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: On the git you still have ClassDefNV, could you push the code where you removed it and it still fails to compile?

**Comment by lietava** [comment]
@shahor02 pushed

**Comment by shahor02** [comment]
@lietava it should not produce such a message with this code, perhaps cleaning of the CTP/simulation dir. in the BUILD/... would help. But you don't need at all to add this class to the library, instead you should directly create an executable, like e.g.
https://github.com/AliceO2Group/AliceO2/blob/1299fa246ed5a13e31cc0625cfb7332c2b082e5a/Detectors/FIT/FT0/simulation/CMakeLists.txt#L34-L41

**Comment by lietava** [comment]
@shahor02 I do not understand this comment:
https://github.com/AliceO2Group/AliceO2/pull/6278#discussion_r644047157

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (lietava) [reply DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: yes, fine. What is motivation ? We want to save memory ? Is it more elegant ?

**Comment by lietava** [comment]
> @lietava it should not produce such a message with this code, perhaps cleaning of the CTP/simulation dir. in the BUILD/... would help. But you don't need at all to add this class to the library, instead you should directly create an executable, like e.g.
> https://github.com/AliceO2Group/AliceO2/blob/1299fa246ed5a13e31cc0625cfb7332c2b082e5a/Detectors/FIT/FT0/simulation/CMakeLists.txt#L34-L41

@shahor02  After rm -r BUILD and rebuild it is ok. I will also proceed in the way you suggested. For my education: What is motivation not having classes in root dictionary ?

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h
Comment (shahor02) [reply DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: The size and serialization speed: if this goes to ccdb rather than to the data, then the size argument is irrelevant, but to send as a DPL message such object or to store it in the file you need to go via root serialization of every data member.
Anyway, this is a minor comment, if the class is for CCDB only.

**Comment by shahor02** [comment]
@lietava the root dict. is needed only if the class should be persistent (including root serialization via DPL) or you want to use it from the root interpreter command line. Since this compiles to a simple executable, the dict. would just generate an unused code, no profit.

**Comment by lietava** [comment]
@shahor02 I added digi2raw.cxx + minor changes you suggested. Please, have a look.

**Review by shahor02 — APPROVED** [review]
looks fine, I've just fixed a misleading option hint in the digi2raw

**Comment by shahor02** [comment]
@lietava I've removed draft status to trigger CI tests. Do you have anything to add here?

**Comment by lietava** [comment]
@shahor02 Thanks. I am adding nothing at the moment.
Please, proceed with review.

**Comment by shahor02** [comment]
@lietava 
Could you correct this ?
````
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/6278/0/DataFormats/Detectors/CTP/src/Configuration.cxx:85:24: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/6278/0/DataFormats/Detectors/CTP/src/Configuration.cxx:87:25: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/6278/0/DataFormats/Detectors/CTP/src/Configuration.cxx:117:20: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/6278/0/DataFormats/Detectors/CTP/src/Configuration.cxx:277:31: error: statement should be inside braces [readability-braces-around-statements]
````
Did you actually try to run digit2raw, does it produce readable file (can be checked by `o2-raw-file-check <rawfilename>`) ?

**Comment by lietava** [comment]
@shahor02 1 I will fix latest on Tuesday. I am off this long weekend.
2 i did not run digi2raw. Marek B can test it. I open issue for him in my rep.

File: Detectors/CTP/simulation/src/digi2raw.cxx
Comment (shahor02) [inline Detectors/CTP/simulation/src/digi2raw.cxx:117]: I guess you should call `m2r.init()` here?

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (shahor02) [inline Detectors/CTP/simulation/src/Digits2Raw.cxx]: The simulation creates 1 single entry with vectors containing all data

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (shahor02) [inline Detectors/CTP/simulation/src/Digits2Raw.cxx]: therefore you should start from 0 (upper bound `digiTree->GetEntries()` does not hurt, but you have just 1 entry).

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (shahor02) [inline Detectors/CTP/simulation/src/Digits2Raw.cxx]: I don't see where do you call `writer.addData(...)`.
BTW, why do you watch orbit change? You just need to call addData for every orbit/bc where you have non-0 payload (and if you call it for 1 link, then this should be done for all. There is a writer-side protection against not doing this, but you better take care of this on your side).

**Review by shahor02 — CHANGES_REQUESTED** [review]

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (lietava) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: @shahor02  Yes, fixing it. 
Orbit: for non zero suppressed data I need to add (bcid, 0) for all bcs which are not in input record.
For the moment one link is zero suppressed and other is not.
 **Question** on emptyHBFMethod:  how can I do it link dependent ?

Comment (shahor02) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: > Orbit: for non zero suppressed data I need to add (bcid, 0) for all bcs which are not in input record.

OK, but it should be done in a single loop, w/o pre-reading the entry. 

> Question on emptyHBFMethod: how can I do it link dependent ?

You receive in the emptyHBFMethod the rdh for which the payload need to be generated, so you have the full info about the link.

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (lietava) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: @shahor02 So can I assume that  I expect one entry, i.e. more than 1 entry is error ?

Comment (shahor02) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: Currently it is 1 entry, but better do not depend on this assumption, can't you loop over all entries?

Comment (lietava) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: yes, ok, just asking. I'll loop.

File: Detectors/CTP/simulation/src/Digits2Raw.cxx
Comment (lietava) [reply Detectors/CTP/simulation/src/Digits2Raw.cxx]: @shahor02 what is PLTrailer ? Do I need it ?

**Comment by lietava** [comment]
HI @shahor02 , I think I did all changes you requested. I added also payload on link 1 (Trigger Classes). Now it produces non empty file.
I am asking Marek to check what is there. Cheers, Roman.

**Comment by lietava** [comment]
@shahor02 Going to move CTP classes calculation from Digits2Raw to Digitizer. 
Related question: we need to connect meaning of CTP inputs with detector inputs (eg V0, T0). Conceptually it can be done either by using input names or also code number - so called signature - we have in run2. Techically I guess I need to access detector config file in database.

**Comment by shahor02** [comment]
@lietava Up to you: the digitizer is supposed to be responsible for the MC digits creation only, MC->raw: for the conversion which will not be done in the standard MC production (there we will use digits directly).
For the inputs: I am not sure if FIT detectors have these signatures, perhaps you should invent one, create a config object on your side and tell to FIT to use them?

**Comment by lietava** [comment]
@shahor02:
1.) I moved detector to input/class translation to Digitizer meaning that Digits2Raw and Raw2Digits will not need CTP config. Difference between raw and digits is only different data packing in CTP case
2.) I will discus with detectors inputs conventions
3.) ROOT and bitset: does it mean we can never change CTP_NINPUTS and CTP_NCLASSES (see Digits.h) as only those are used in the class which is used by ROOT ? ( I mean there is also CTP_MAXTRIGINPPERDET).
4.) What next : merge or should I add Raw2Digit ?

**Comment by shahor02** [comment]
> 3.) ROOT and bitset: does it mean we can never change CTP_NINPUTS and CTP_NCLASSES (see Digits.h) as only those are used in the class which is used by ROOT ? ( I mean there is also CTP_MAXTRIGINPPERDET).

Yes, ROOT schema evolution does not support template parameters change, and since bitset<N1> and bitset<N2> are different classes, if you change the constant, you cannot read old data.

For that reason in the detector mask (https://github.com/AliceO2Group/AliceO2/blob/aa1308a84e367aa334e2fb51c7f50cd036af77ef/DataFormats/Detectors/Common/include/DetectorsCommonDataFormats/DetID.h#L90) we are using fixed 32-slot mask although we have only 17 (or 20 with future upgrades) detectors. Note that we are using custom implementation of David, https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/Utils/GPUCommonBitSet.h which takes only 32 bits and cannot sore more (the std::bitset takes at least 64 bits)

> 4.) What next : merge or should I add Raw2Digit ?
We cannot use raw w/o decoder, but I would merge this once the tests are passed, to avoid divergences.

**Comment by lietava** [comment]
@shahor02 ok, thanks, waiting for merge finished.  Give me examples on decoder, please.

**Comment by lietava** [comment]
@shahor02 Do you still have request or is it previous ?

**Comment by shahor02** [comment]
Merging, the mask problem, if needed, can be addressed in other PR. For the decoding, e.g., https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/CPV/workflow/src/RawToDigitConverterSpec.cxx, which is a part of https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/CPV/workflow/src/cpv-reco-workflow.cxx.

**Comment by lietava** [comment]
@shahor02 : what is 'the mask problem' ?

**Comment by shahor02** [comment]
The bitsets you were asking about. If you think the template parameters may change, better to reserve at once sufficient number of bits, even if now you don't use them.

**Pull request merged**
